src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -0,0 +1,1653 @@
+  // Note: for shortness, just print out the first 32. That should
+  // cover most of the useful ones, apart from realtime signals.
+  for (int sig = 1; sig <= NUM_IMPORTANT_SIGS; sig++) {
+    const int rc = sigismember(set, sig);
+    if (rc == -1 && errno == EINVAL) {
+      buffer[sig-1] = '?';
+    } else {
+      buffer[sig-1] = rc == 0 ? '0' : '1';
+    }
+  }
+  buffer[NUM_IMPORTANT_SIGS] = 0;
+  return buffer;
+}
+
+// Prints one-line description of a signal set.
+static void print_signal_set_short(outputStream* st, const sigset_t* set) {
+  char buf[NUM_IMPORTANT_SIGS + 1];
+  describe_signal_set_short(set, buf, sizeof(buf));
+  st->print("%s", buf);
+}
+
+void PosixSignals::print_signal_handler(outputStream* st, int sig,
+                                 char* buf, size_t buflen) {
+  struct sigaction sa;
+  sigaction(sig, NULL, &sa);
+
+  // See comment for SIGNIFICANT_SIGNAL_MASK define
+  sa.sa_flags &= SIGNIFICANT_SIGNAL_MASK;
+
+  st->print("%s: ", os::exception_name(sig, buf, buflen));
+
+  address handler = (sa.sa_flags & SA_SIGINFO)
+    ? CAST_FROM_FN_PTR(address, sa.sa_sigaction)
+    : CAST_FROM_FN_PTR(address, sa.sa_handler);
+
+  if (handler == CAST_FROM_FN_PTR(address, SIG_DFL)) {
+    st->print("SIG_DFL");
+  } else if (handler == CAST_FROM_FN_PTR(address, SIG_IGN)) {
+    st->print("SIG_IGN");
+  } else {
+    st->print("[%s]", get_signal_handler_name(handler, buf, buflen));
+  }
+
+  st->print(", sa_mask[0]=");
+  print_signal_set_short(st, &sa.sa_mask);
+
+  address rh = VMError::get_resetted_sighandler(sig);
+  // May be, handler was resetted by VMError?
+  if (rh != NULL) {
+    handler = rh;
+    sa.sa_flags = VMError::get_resetted_sigflags(sig) & SIGNIFICANT_SIGNAL_MASK;
+  }
+
+  // Print textual representation of sa_flags.
+  st->print(", sa_flags=");
+  print_sa_flags(st, sa.sa_flags);
+
+  // Check: is it our handler?
+  if (handler == CAST_FROM_FN_PTR(address, (sa_sigaction_t)javaSignalHandler) ||
+      handler == CAST_FROM_FN_PTR(address, (sa_sigaction_t)SR_handler)) {
+    // It is our signal handler
+    // check for flags, reset system-used one!
+    if ((int)sa.sa_flags != get_our_sigflags(sig)) {
+      st->print(
+                ", flags was changed from " PTR32_FORMAT ", consider using jsig library",
+                get_our_sigflags(sig));
+    }
+  }
+  st->cr();
+}
+
+bool PosixSignals::is_sig_ignored(int sig) {
+  struct sigaction oact;
+  sigaction(sig, (struct sigaction*)NULL, &oact);
+  void* ohlr = oact.sa_sigaction ? CAST_FROM_FN_PTR(void*,  oact.sa_sigaction)
+                                 : CAST_FROM_FN_PTR(void*,  oact.sa_handler);
+  if (ohlr == CAST_FROM_FN_PTR(void*, SIG_IGN)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+int PosixSignals::unblock_thread_signal_mask(const sigset_t *set) {
+  return pthread_sigmask(SIG_UNBLOCK, set, NULL);
+}
+
+address PosixSignals::ucontext_get_pc(const ucontext_t* ctx) {
+#if defined(AIX)
```

Comment:
Again there must be a better way to do this dispatch. If the target were os::ucontext_get_pc, defined is os<os>.cpp then we would link to the current platforms version.

Reply:
Will be addressed in JDK-8253742